tests/lean/run/beta_reduce_mixfix.lean

```diff
@@ -0,0 +1,2 @@
+local infixr ` $> `:100  := λ a b, b <$ a
+example (m : Type → Type) [monad m] : m ℕ := pure 0 $> 1
```

Comment:
The problem is that elaboration fails without the beta reduction

Reply:
I think the main issue is the way we elaborate `(λ ...) (pure 0) 1`. We don't use information about the type of `pure 0` and `1` when elaborating `(λ ...)`. Actually, we elaborate `(λ ...)` without any expected type information. Note that we usually propagate typing information from left to right. That is, we elaborate `f a` by first elaborating `f` and then using information about type of `f` to elaborate `a`.

The `head_beta_reduce` trick above may backfire in the future and produce unintended behavior. Example: the user is using a lambda to create a let-like expression.
 
I think the following should work
1- We define a constant `map_const_rev` as ` λ a b, b <$ a`
2- We define ` $> ` as notation for `map_const_rev`
Benefit: we preserve the user intent. That is, `a $> b` will be pretty printed as `a $> b` instead of `b <$ a`.

Reply:
> I think the main issue is the way we elaborate (λ ...) (pure 0) 1.

Do you propose we should change that way? I encountered a related issue where I have to write `has_map.map id` instead of `(id <$>)` because the latter introduces an unnecessary abstraction that again interferes with elaboration.

> The head_beta_reduce trick above may backfire in the future and produce unintended behavior.

We could restrict it to do either zero or exactly two reductions, that should hopefully not surprise anyone.

> We define a constant map_const_rev as λ a b, b <$ a

I really would like to avoid that - I already find it bothersome that I have to unfold `ge` to apply `le` simp lemmas (though an `add_key_equivalence ge le` may fix that?).
The new conservative pretty printer should help in preserving the user's intent. As one more data point, Isabelle does not distinguish between `<` and `>` at all and prints them both as `<`.

Reply:
> Do you propose we should change that way? 

No, I fear it would make the elaborator even more complicated and unpredictable.
If we decide to add new elaboration strategies, we should at least provide annotations and/or attributes for letting users choose the one they want.

> I encountered a related issue where I have to write has_map.map id instead of (id <$>) because the latter introduces an unnecessary abstraction that again interferes with elaboration.

In both cases, the problem was introduced by an auxiliary lambda abstraction.
One option is to have an annotation for instructing the elaborator to apply beta reduction.
The annotation can be something simple like an identity function named `beta`.
Then, during elaboration `((beta f) a b)` is reduced to `head_beta_reduce (f a b)`. 

> We could restrict it to do either zero or exactly two reductions, that should hopefully not surprise anyone.

I think we need to let the user decide when to use the beta reduction. This is why I suggested the `beta` trick above. Otherwise, I bet someone will be surprised, and our explanation will sound weird: we do zero or exactly two reductions.

> I really would like to avoid that - I already find it bothersome that I have to unfold ge to apply le simp lemmas (though an add_key_equivalence ge le may fix that?).

I don't see a problem here. You can add a `simp` lemma that reduces `ge` into `le`. 
This kind of lemma would be a good candidate for the new "normalizing" `simp` set we discussed last week. Other users may decide to add this lemma to the main `simp` set. 

> The new conservative pretty printer should help in preserving the user's intent.

How would we preserve the user's intent after we apply tactics such as `simp`? It sound tricky.

> As one more data point, Isabelle does not distinguish between < and > at all and prints them both as <.

As I said above, we can easily simulate this behavior by adding a new lemma to a `simp` set.
We are providing more options to users: we can keep `>` or reduce it. 
I don't see how this can be a bad thing.
BTW, Coq doesn't print them both as `<`.

Reply:
Okay, I think I should get used to the Coq way. I'll introduce `map_const_rev` then.